components-api/src/main/scala/pl/touk/nussknacker/engine/api/generics/TypingFunction.scala

```diff
@@ -4,35 +4,24 @@ import cats.data.ValidatedNel
 import pl.touk.nussknacker.engine.api.typed.typing.TypingResult
 
 /**
- * Base for creating custom classes that extend methods signature.
+ * Class representing additional information about methods type.
  *
  * <p>
  * Deriving classes must be declared as non-anonymous class or case class,
  * they have to be static or declared at top level, and they must have
  * parameterless constructor. More precisely, they must be instantiable
  * using:
  * <pre>typeFunctionClass.getDeclaredConstructor().newInstance()</pre>
- *
- * <p>
- * Constructor of deriving class will be called every time appropriate
- * method is validated, so it should not do any unnecessary computations.
  */
 abstract class TypingFunction {
   /**
-   * Approximation of types of parameters that can be accepted
-   * by method. Used for displaying information about method on FE
-   * and generating error messages. Defaults to types that can be
-   * extracted from methods signature if it is not specified.
-   */
-  def staticParameters: Option[ParameterList] = None
-
-  /**
-   * Approximation of return type of method. Used for displaying
-   * information about method on FE and for method suggestions.
-   * Defaults to type extracted from methods signature if it is
-   * not specified.
+   * List of possible combinations of parameters and result types that
+   * this function can accept. Must be at least as specific as types that
+   * can be derived from associated method and less specific than
+   * computeResultType.
+   * Defaults to type of associated method if empty list is provided.
    */
-  def staticResult: Option[TypingResult] = None
+  def signatures: List[MethodTypeInfo] = Nil
```

Comment:
WDYT about Option[Nel]? will be more obvious that for None, it will be generated.